### DISCUSSION TITLE
ViewController is presented for Optional Destination Reducer case payload

### DISCUSSION THREAD

**Discussion by BastianKusserow** [pointfreeco/swift-composable-architecture#3388] [General]
### Description

Hey!

I just encountered a unexpected behaviour in TCA in conjunction with swift-navigation. Whenever you have a Destination reducer that presents non TCA features, if the associated value of that case is optional, the presentation tools immediately present those features / invoke the presentation closure of swift-navigation. Additionaly, whenever I try to present the feature again, after dismissing the first screen, nothing happens.

A reproduction example of the issue is attached.

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

I'd expect that the feature is only displayed when the destination state actual matches the case.

### Actual behavior

the feature is displayed immediately and cannot be presented again.

### Reproducing project

[UIKitNavigation.zip](https://github.com/user-attachments/files/17010934/UIKitNavigation.zip)

### The Composable Architecture version information

1.15.0

### Destination operating system

iOS 17

### Xcode version information

15.4

**Comment by mbrandonw** [comment]
Hi @BastianKusserow, this is the expected behavior. The optionality of `Destination.State?` is what controls the presentation/dismissal of the view, regardless of what the payload is. It is completely valid to present a sheet with a piece of optional state. We do not do extra work to squash the double optional as that would be a little too magical and probably unexpected in some cases.

And as for the second issue you mentioned, your project doesn't actually compile for me, but that behavior seems to work just fine when I run the various case studies from the Swift Navigation repo. If you can create a repro project that compiles I can take a look.

Since this isn't an issue with the library I am going to convert it to a discussion. Please feel free to continue the conversation over there!

**Reply by BastianKusserow** [reply]
Hey @mbrandonw, thanks for taking a look into the issue. Sorry that the project did not work, could you try again with [UIKitNavigationOptional.zip](https://github.com/user-attachments/files/17013129/UIKitNavigationOptional.zip)?

I get that the presentation is driven of the optionality of `Destination.State?`, but what I don't get is why the modal is immediately presented when starting the app **although** the destination is still nil in that point in time. Also, if I set the destination to the `case rating(nil)` or `case rating(1)`, the sheet is not presented anymore. My expectation would be that, when I scope to  `$store.scope(state: \.destination?.rating, action: \.destination.rating)` the case of `state.destination` would at least have to be `.rating`. Sorry if I don't get the point here, but for me the behaviour was quite unexpected.

**Reply by stephencelis** [reply]
@BastianKusserow This sadly seems to be a Swift bug around optional-chained key paths and nested optionals. We opened an issue here:

https://github.com/swiftlang/swift/issues/76510

To avoid it, we would recommend not storing optionals directly in the enum payload to avoid the nested optional access, and instead wrap the optional in a struct.

**Reply by BastianKusserow** [reply]
Ah, thanks for clarifying. Does that mean the project that I sent you should work as expected?

In the meantime I boxed the optional in a struct like you've mentioned and it works as expected, thanks!
